File: 连接格点.py
import sys
input = lambda: sys.stdin.readline().strip()

class DSU: # 并查集
    def __init__(self, n):
        self.pre = list(range(n + 1))
        self.siz = [1 for _ in range(n + 1)]
    
    def find(self, x):
        x, xcopy = self.pre[x], x
        while x != self.pre[x]:
            x = self.pre[x]
        while xcopy != x:
            self.pre[xcopy], xcopy = x, self.pre[xcopy]
        return x
    
    def same(self, x, y):
        return self.find(x) == self.find(y)

    def merge(self, x, y):
        x, y = self.find(x), self.find(y)
        if x == y:
            return False
        self.siz[x] += self.siz[y]
        self.pre[y] = x
        return True
    
    def size(self, x):
        return self.siz[self.find(x)]
    
n, m = map(int, input().split())
ans = 0
dsu = DSU(n * m + 1)
for line in sys.stdin:
    x1, y1, x2, y2 = map(int, line.split())
    u, v = (x1 - 1) * m + y1, (x2 - 1) * m + y2
    dsu.merge(u, v)
for i in range(1, m + 1):
    for j in range(1, n):
        u, v = (j - 1) * m + i, j * m + i
        if dsu.merge(u, v):
            ans += 1

for i in range(1, n + 1):
    for j in range(1, m):
        u, v = (i - 1) * m + j, (i - 1) * m + j + 1
        if dsu.merge(u, v):
            ans += 2
print(ans)

File: 连接格点.cpp
#include <bits/stdc++.h>
using namespace std;

#define ff first
#define ss second
#define int long long

struct DSU { // 并查集
    vector<int> pre, siz;
    
    DSU() {}
    DSU(int n) {
        pre.resize(n);
        std::iota(pre.begin(), pre.end(), 0);
        siz.assign(n, 1);
    }
    
    int find(int x) {
        if (pre[x] == x) {
            return x;
        }
        return pre[x] = find(pre[x]);
    }
    
    bool same(int x, int y) {
        return find(x) == find(y);
    }
    
    bool merge(int x, int y) {
        x = find(x);
        y = find(y);
        if (x == y) {
            return false;
        }
        siz[x] += siz[y];
        pre[y] = x;
        return true;
    }
    
    int size(int x) {
        return siz[find(x)];
    }
};

signed main() {
    int n, m, ans = 0;
    cin >> n >> m;
    DSU dsu(n * m + 1);
    int x1, x2, y1, y2;
    while (cin >> x1 >> y1 >> x2 >> y2) {
        int u = (x1 - 1) * m + y1, v = (x2 - 1) * m + y2;
        dsu.merge(u, v);
    }
    
    // 由于竖向花费最小，所以先判断竖向
    for (int i = 1; i <= m; ++i) {
        for (int j = 1; j < n; ++j) {
            int u = (j - 1) * m + i, v = j * m + i;
            if (dsu.merge(u, v)) {
                ans += 1;
            }
        }
    }
    // 横向合并
    for (int i = 1; i <= n; ++i) {
        for (int j = 1; j < m; ++j) {
            int u = (i - 1) * m + j, v = (i - 1) * m + j + 1;
            if (dsu.merge(u, v)) {
                ans += 2; 
            }
        }
    }
    cout << ans << endl;
    return 0;
}
